File: python/lib/recommender.py
"""
Two-Tower Recommendation Engine for Connect3

This implements the full recommendation pipeline:
1. Load event embeddings
2. Compute user embedding
3. Retrieve top-N candidates via vector similarity
4. Apply business rules (recency, diversity, etc.)
5. Return final recommendations
"""
import re
import json
from datetime import datetime
from typing import List, Dict, Any, Optional, Set
from dataclasses import dataclass, field

from .supabase_client import supabase
from .embeddings import embedding_service, CONNECT3_CATEGORIES
from .vector_index import VectorIndex, get_event_index, SearchResult
from .config import RECOMMENDATION_CONFIG


@dataclass
class RecommendedEvent:
    event_id: str
    title: str
    caption: str
    timestamp: str
    permalink: str
    media_url: str
    category: Optional[str]
    similarity_score: float
    recency_score: float
    final_score: float
    reason: str = ""  # Why this event is recommended


@dataclass
class RecommendationConfig:
    top_k: int = 10  # Number of final recommendations
    candidate_multiplier: int = 3  # Retrieve top_k * this for re-ranking
    recency_weight: float = 0.3  # Weight for recency (0-1)
    similarity_weight: float = 0.7  # Weight for similarity (0-1)
    diversity_penalty: float = 0.1  # Penalty for same-category events
    max_days_old: int = 60  # Filter out events older than this


DEFAULT_CONFIG = RecommendationConfig()


class TwoTowerRecommender:
    """Two-tower recommendation engine using embeddings"""
    
    def __init__(self, config: Optional[RecommendationConfig] = None):
        self.event_index = get_event_index()
        self.config = config or DEFAULT_CONFIG
    
    def load_event_index(self) -> None:
        """Load event embeddings into the vector index"""
        # Fetch all event embeddings from database
        result = supabase.table("event_embeddings").select("event_id, embedding, category, created_at").execute()
        event_embeddings = result.data or []
        
        if not event_embeddings:
            print("No event embeddings found in database")
            return
        
        # Load into vector index
        self.event_index.clear()
        
        for event in event_embeddings:
            embedding = (
                json.loads(event["embedding"])
                if isinstance(event["embedding"], str)
                else event["embedding"]
            )
            
            self.event_index.add(event["event_id"], embedding, {
                "category": event.get("category"),
                "created_at": event.get("created_at")
            })
        
        print(f"Loaded {len(event_embeddings)} events into vector index")
    
    def get_recommendations(self, user_id: str) -> List[RecommendedEvent]:
        """Get recommendations for a user"""
        # Step 1: Get user embedding
        user_embedding = embedding_service.embed_user(user_id)
        
        # Step 2: Get user's past interactions to exclude
        result = supabase.table("feedback_logs").select("event_id").eq("user_id", user_id).execute()
        past_interactions = result.data or []
        exclude_ids = set(i["event_id"] for i in past_interactions)
        
        # Step 3: Retrieve candidates via vector search
        num_candidates = self.config.top_k * self.config.candidate_multiplier
        candidates = self.event_index.search(
            user_embedding.embedding,
            num_candidates,
            exclude_ids
        )
        
        if not candidates:
            return []
        
        # Step 4: Fetch full event data for candidates
        candidate_ids = [c.id for c in candidates]
        result = supabase.table("events").select("*").in_("id", candidate_ids).execute()
        events = result.data or []
        
        if not events:
            return []
        
        # Step 5: Apply business rules and re-rank
        ranked_events = self._apply_business_rules(candidates, events)
        
        # Step 6: Add recommendation reasons
        return [
            RecommendedEvent(
                event_id=e["event_id"],
                title=e["title"],
                caption=e["caption"],
                timestamp=e["timestamp"],
                permalink=e["permalink"],
                media_url=e["media_url"],
                category=e["category"],
                similarity_score=e["similarity_score"],
                recency_score=e["recency_score"],
                final_score=e["final_score"],
                reason=self._generate_recommendation_reason(e)
            )
            for e in ranked_events[:self.config.top_k]
        ]
    
    def _apply_business_rules(
        self,
        candidates: List[SearchResult],
        events: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Apply business rules for re-ranking"""
        now = datetime.now()
        scored_events = []
        
        # Create lookup map for events
        event_map = {e["id"]: e for e in events}
        
        # Track category counts for diversity
        category_counts: Dict[str, int] = {}
        
        for candidate in candidates:
            event = event_map.get(candidate.id)
            if not event:
                continue
            
            # Calculate recency score (newer = higher)
            event_date_str = event.get("timestamp") or event.get("event_date")
            if event_date_str:
                try:
                    event_date = datetime.fromisoformat(event_date_str.replace("Z", "+00:00"))
                    days_old = abs((event_date - now).days)
                except:
                    days_old = 0
            else:
                days_old = 0
            
            # Filter out old events
            if days_old > self.config.max_days_old:
                continue
            
            # Recency score: 1 for today, decreasing with age
            recency_score = max(0, 1 - (days_old / self.config.max_days_old))
            
            # Diversity penalty
            category = event.get("category") or "unknown"
            category_count = category_counts.get(category, 0)
            diversity_penalty = category_count * self.config.diversity_penalty
            category_counts[category] = category_count + 1
            
            # Final score
            final_score = (
                (candidate.score * self.config.similarity_weight) +
                (recency_score * self.config.recency_weight) -
                diversity_penalty
            )
            
            scored_events.append({
                "event_id": event["id"],
                "title": event.get("title") or self._extract_title(event.get("caption", "")),
                "caption": event.get("caption") or event.get("description", ""),
                "timestamp": event.get("timestamp") or event.get("event_date", ""),
                "permalink": event.get("permalink") or event.get("source_url", ""),
                "media_url": event.get("media_url", ""),
                "category": event.get("category"),
                "similarity_score": candidate.score,
                "recency_score": recency_score,
                "final_score": final_score
            })
        
        # Sort by final score
        scored_events.sort(key=lambda e: e["final_score"], reverse=True)
        
        return scored_events
    
    def _extract_title(self, caption: str) -> str:
        """Extract title from caption (first line or first sentence)"""
        if not caption:
            return "Event"
        
        # Get first line
        first_line = caption.split("\n")[0]
        
        # Clean emojis and truncate
        emoji_pattern = re.compile(
            "["
            "\U0001F600-\U0001F64F"
            "\U0001F300-\U0001F5FF"
            "\U0001F680-\U0001F6FF"
            "]+",
            flags=re.UNICODE
        )
        cleaned = emoji_pattern.sub("", first_line).strip()
        
        return cleaned[:100] + "..." if len(cleaned) > 100 else cleaned
    
    def _generate_recommendation_reason(self, event: Dict[str, Any]) -> str:
        """Generate human-readable recommendation reason"""
        reasons = []
        
        # High similarity
        sim_score = event.get("similarity_score", 0)
        if sim_score > 0.8:
            reasons.append("Matches your interests very well")
        elif sim_score > 0.6:
            reasons.append("Related to topics you enjoy")
        
        # Category-based reason
        category = event.get("category")
        if category:
            category_names = {
                "tech_workshop": "tech and coding workshops",
                "career_networking": "career and networking events",
                "hackathon": "hackathons and competitions",
                "social_event": "social activities",
                "academic_revision": "academic support",
                "recruitment": "club activities",
                "industry_talk": "industry insights",
                "sports_recreation": "sports and recreation",
                "entrepreneurship": "entrepreneurship",
                "community_service": "community events"
            }
            category_name = category_names.get(category, category)
            reasons.append(f"You've shown interest in {category_name}")
        
        # Recent event
        if event.get("recency_score", 0) > 0.8:
            reasons.append("Happening soon")
        
        return " • ".join(reasons) if reasons else "Recommended for you"
    
    def get_batch_recommendations(
        self, 
        user_ids: List[str]
    ) -> Dict[str, List[RecommendedEvent]]:
        """Get recommendations for all users (batch mode for email sending)"""
        results = {}
        
        for user_id in user_ids:
            try:
                recommendations = self.get_recommendations(user_id)
                results[user_id] = recommendations
            except Exception as e:
                print(f"Failed to get recommendations for user {user_id}: {e}")
                results[user_id] = []
        
        return results


# Singleton instance
recommender = TwoTowerRecommender()

File: src/lib/recommender.ts
/**
 * Two-Tower Recommendation Engine for Connect3
 * 
 * This implements the full recommendation pipeline:
 * 1. Load event embeddings
 * 2. Compute user embedding
 * 3. Retrieve top-N candidates via vector similarity
 * 4. Apply business rules (recency, diversity, etc.)
 * 5. Return final recommendations
 */

import { supabase } from './supabase';
import { embeddingService, CONNECT3_CATEGORIES, Connect3Category } from './embeddings';
import { VectorIndex, getEventIndex, SearchResult } from './vector-index';
import { differenceInDays } from 'date-fns';

export interface RecommendedEvent {
  event_id: string;
  title: string;
  caption: string;
  timestamp: string;
  permalink: string;
  media_url: string;
  category: Connect3Category | null;
  similarity_score: number;
  recency_score: number;
  final_score: number;
  reason: string; // Why this event is recommended
}

export interface RecommendationConfig {
  topK: number;               // Number of final recommendations
  candidateMultiplier: number; // Retrieve topK * this for re-ranking
  recencyWeight: number;      // Weight for recency (0-1)
  similarityWeight: number;   // Weight for similarity (0-1)
  diversityPenalty: number;   // Penalty for same-category events
  maxDaysOld: number;         // Filter out events older than this
}

const DEFAULT_CONFIG: RecommendationConfig = {
  topK: 10,
  candidateMultiplier: 3,
  recencyWeight: 0.3,
  similarityWeight: 0.7,
  diversityPenalty: 0.1,
  maxDaysOld: 60,
};

export class TwoTowerRecommender {
  private eventIndex: VectorIndex;
  private config: RecommendationConfig;

  constructor(config: Partial<RecommendationConfig> = {}) {
    this.eventIndex = getEventIndex();
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Load event embeddings into the vector index
   */
  async loadEventIndex(): Promise<void> {
    // Fetch all event embeddings from database
    const { data: eventEmbeddings, error } = await supabase
      .from('event_embeddings')
      .select('event_id, embedding, category, created_at');

    if (error) {
      throw new Error(`Failed to load event embeddings: ${error.message}`);
    }

    if (!eventEmbeddings || eventEmbeddings.length === 0) {
      console.log('No event embeddings found in database');
      return;
    }

    // Load into vector index
    this.eventIndex.clear();
    
    for (const event of eventEmbeddings) {
      const embedding = typeof event.embedding === 'string'
        ? JSON.parse(event.embedding)
        : event.embedding;

      this.eventIndex.add(event.event_id, embedding, {
        category: event.category,
        created_at: event.created_at,
      });
    }

    console.log(`Loaded ${eventEmbeddings.length} events into vector index`);
  }

  /**
   * Get recommendations for a user
   */
  async getRecommendations(userId: string): Promise<RecommendedEvent[]> {
    // Step 1: Get user embedding
    const userEmbedding = await embeddingService.embedUser(userId);

    // Step 2: Get user's past interactions to exclude
    const { data: pastInteractions } = await supabase
      .from('feedback_logs')
      .select('event_id')
      .eq('user_id', userId);

    const excludeIds = new Set(pastInteractions?.map(i => i.event_id) || []);

    // Step 3: Retrieve candidates via vector search
    const numCandidates = this.config.topK * this.config.candidateMultiplier;
    const candidates = this.eventIndex.search(
      userEmbedding.embedding,
      numCandidates,
      excludeIds
    );

    if (candidates.length === 0) {
      return [];
    }

    // Step 4: Fetch full event data for candidates
    const candidateIds = candidates.map(c => c.id);
    const { data: events } = await supabase
      .from('events')
      .select('*')
      .in('id', candidateIds);

    if (!events) return [];

    // Step 5: Apply business rules and re-rank
    const rankedEvents = this.applyBusinessRules(candidates, events);

    // Step 6: Add recommendation reasons
    return rankedEvents.slice(0, this.config.topK).map(event => ({
      ...event,
      reason: this.generateRecommendationReason(event),
    }));
  }

  /**
   * Apply business rules for re-ranking
   */
  private applyBusinessRules(
    candidates: SearchResult[],
    events: any[]
  ): Omit<RecommendedEvent, 'reason'>[] {
    const now = new Date();
    const scoredEvents: Omit<RecommendedEvent, 'reason'>[] = [];

    // Create lookup map for events
    const eventMap = new Map(events.map(e => [e.id, e]));

    // Track category counts for diversity
    const categoryCounts = new Map<string, number>();

    for (const candidate of candidates) {
      const event = eventMap.get(candidate.id);
      if (!event) continue;

      // Calculate recency score (newer = higher)
      const eventDate = new Date(event.timestamp || event.event_date);
      const daysOld = Math.abs(differenceInDays(eventDate, now));

      // Filter out old events
      if (daysOld > this.config.maxDaysOld) continue;

      // Recency score: 1 for today, decreasing with age
      const recencyScore = Math.max(0, 1 - (daysOld / this.config.maxDaysOld));

      // Diversity penalty
      const category = event.category || 'unknown';
      const categoryCount = categoryCounts.get(category) || 0;
      const diversityPenalty = categoryCount * this.config.diversityPenalty;
      categoryCounts.set(category, categoryCount + 1);

      // Final score
      const finalScore = 
        (candidate.score * this.config.similarityWeight) +
        (recencyScore * this.config.recencyWeight) -
        diversityPenalty;

      scoredEvents.push({
        event_id: event.id,
        title: event.title || this.extractTitle(event.caption),
        caption: event.caption || event.description,
        timestamp: event.timestamp || event.event_date,
        permalink: event.permalink || event.source_url,
        media_url: event.media_url || '',
        category: event.category,
        similarity_score: candidate.score,
        recency_score: recencyScore,
        final_score: finalScore,
      });
    }

    // Sort by final score
    scoredEvents.sort((a, b) => b.final_score - a.final_score);

    return scoredEvents;
  }

  /**
   * Extract title from caption (first line or first sentence)
   */
  private extractTitle(caption: string): string {
    if (!caption) return 'Event';
    
    // Get first line
    const firstLine = caption.split('\n')[0];
    
    // Clean emojis and truncate
    const cleaned = firstLine
      .replace(/[\u{1F600}-\u{1F64F}]/gu, '')
      .replace(/[\u{1F300}-\u{1F5FF}]/gu, '')
      .replace(/[\u{1F680}-\u{1F6FF}]/gu, '')
      .trim();

    return cleaned.length > 100 ? cleaned.substring(0, 100) + '...' : cleaned;
  }

  /**
   * Generate human-readable recommendation reason
   */
  private generateRecommendationReason(event: Omit<RecommendedEvent, 'reason'>): string {
    const reasons: string[] = [];

    // High similarity
    if (event.similarity_score > 0.8) {
      reasons.push('Matches your interests very well');
    } else if (event.similarity_score > 0.6) {
      reasons.push('Related to topics you enjoy');
    }

    // Category-based reason
    if (event.category) {
      const categoryNames: Record<string, string> = {
        'tech_workshop': 'tech and coding workshops',
        'career_networking': 'career and networking events',
        'hackathon': 'hackathons and competitions',
        'social_event': 'social activities',
        'academic_revision': 'academic support',
        'recruitment': 'club activities',
        'industry_talk': 'industry insights',
        'sports_recreation': 'sports and recreation',
        'entrepreneurship': 'entrepreneurship',
        'community_service': 'community events',
      };
      
      const categoryName = categoryNames[event.category] || event.category;
      reasons.push(`You've shown interest in ${categoryName}`);
    }

    // Recent event
    if (event.recency_score > 0.8) {
      reasons.push('Happening soon');
    }

    return reasons.length > 0 ? reasons.join(' • ') : 'Recommended for you';
  }

  /**
   * Get recommendations for all users (batch mode for email sending)
   */
  async getBatchRecommendations(
    userIds: string[]
  ): Promise<Map<string, RecommendedEvent[]>> {
    const results = new Map<string, RecommendedEvent[]>();

    // Process in parallel with concurrency limit
    const batchSize = 10;
    
    for (let i = 0; i < userIds.length; i += batchSize) {
      const batch = userIds.slice(i, i + batchSize);
      
      const batchResults = await Promise.all(
        batch.map(async (userId) => {
          try {
            const recommendations = await this.getRecommendations(userId);
            return { userId, recommendations };
          } catch (error) {
            console.error(`Failed to get recommendations for user ${userId}:`, error);
            return { userId, recommendations: [] };
          }
        })
      );

      for (const { userId, recommendations } of batchResults) {
        results.set(userId, recommendations);
      }
    }

    return results;
  }
}

export const recommender = new TwoTowerRecommender();
